Make font size dropdown more accessible.

alexstine opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

The overall change in this pull request will allow screen readers to see the currently selected font size without having to open the menu. This message is communicated via aria-describedby. Since the aria-label overrides the button text, I had to provide a way for screen readers to see the currently selected font size value.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

I tested this using NVDA, JAWS, and Voiceover. Browsers used include Firefox, Google Chrome, and Safari.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

I guess you could consider this a Bug fix.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

alexstine committed:
Make font size dropdown more accessible.

alexstine commented:
I believe this is mostly ready for review. This brings a good improvement for the font size component as well as support for other components that may use this custom select box. This PR was primarily created for the font size fixes but there were a couple components involved that needed adjusting.

jorgefilipecosta commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -120,6 +121,7 @@ export default function CustomSelectControl( {
 					'aria-labelledby': undefined,
 					className: 'components-custom-select-control__button',
 					isSmall: true,
+					describedBy: describedBy ? describedBy : undefined,
```
Minor: I guess here we could simply pass `describedBy,` instead of `describedBy: describedBy ? describedBy : undefined,. When the value is on passed describedBy is already undefined if something is passed we would simply pass it down to the button.

alexstine replied:
I am new to React, I thought since it was a passed in property I had to check it. Anyway, corrected.

jorgefilipecosta commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -56,6 +56,7 @@ export default function CustomSelectControl( {
 	className,
 	hideLabelFromVision,
 	label,
+	describedBy,
```
I guess we should document the new property on packages/components/src/custom-select-control/README.md.

alexstine replied:
Done. Do you think it is clear enough?

alexstine replied:
Made another update to usage.

jorgefilipecosta reviewed (approved):
The code change looks good to me and works well on my tests 👍

tellthemachines reviewed (approved):
Tested with VoiceOver/Safari and NVDA/Firefox and this is working really well ✅

alexstine committed:
Address feedback.

alexstine committed:
Merge branch 'trunk' of https://github.com/wordpress/gutenberg into update/font-size-select-accessibility

ntsekouras reviewed (commented):
Thanks for the contribution @alexstine !

I was wondering if we should make this built in in the `CustomSelectControl` component. There are other usages of this component (check font weight, etc...) and we could use the a11y enhancement for all of them. For example we could have something like:

```
const describedBy = `Currently selected ${ label?.toLowerCase() }: ${
	selectedItem?.name
}`;
```

We could even have the new `describedBy` prop and if not provided, use something like the above snippet. What do you think?

youknowriad commented on packages/components/src/font-size-picker/index.js:
```diff
@@ -92,6 +92,10 @@ function FontSizePicker(
 
 	const selectedFontSizeSlug = getSelectValueFromFontSize( fontSizes, value );
 
+	const currentFontSizeSR = `Currently selected font size: ${
```
This probably need to be translated properly?

alexstine replied:
I think it should be good now.

carolinan commented:
Tested with NVDA and chrome on Windows, works well. +1 for using it in the component and solving the translation.

alexstine committed:
Add a fallback if describedBy is not passed. Add translation.

alexstine committed:
Merge branch 'trunk' of https://github.com/wordpress/gutenberg into update/font-size-select-accessibility

alexstine commented:
I made a few updates to the PR.

1. I added a fallback if there was no describedBy text passed in.
2. I decided it was better to allow describedBy to have a custom message passed in. See how descriptive I was able to make font size? Ideally, developers will follow suit and avoid one size fits all. However, if nothing is provided, number 1 takes care of that now.
3. I added sprintf, translation, and translation comments.

Please let me know if an issue is spotted.

@ntsekouras Thanks for the suggestion. As you can see, I adapted your idea a bit, but I think it will work fine this way.

ntsekouras commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -77,6 +79,12 @@ export default function CustomSelectControl( {
 
+	const describedByBuild = describedBy
+		? describedBy
+		: sprintf(
```
This makes linter unhappy 😄

alexstine replied:
The auto-format keeps changing it back. How should it look?

alexstine replied:
I think I got it now.

ntsekouras commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -77,6 +79,12 @@ export default function CustomSelectControl( {
 		stateReducer,
 	} );
 
+	const describedByBuild = describedBy
```
Nitpick: I'm not sure if we can improve the name of this var.. It's easy to find its usage and understand it, though maybe a name like `controlDescribedBy` or something like that 🤔 would be better?

alexstine replied:
Done.

ntsekouras reviewed (commented):
Thanks for your work on this @alexstine ! Great work 💯 

This  just needs the fixing of the linter issue. Maybe a better name to the var as well, but I don't consider this a blocker.

alexstine committed:
Use better variable name. Try to fix linter error.

alexstine committed:
Fix linter errors.

ntsekouras committed:
fix linting issue

ntsekouras reviewed (approved):
Great work here, thanks! 💯
